lib/events/filelog.go

```diff
@@ -202,31 +202,66 @@ func (l *FileLog) SearchEvents(fromUTC, toUTC time.Time, namespace string, event
 	if limit > defaults.EventsMaxIterationLimit {
 		return nil, "", trace.BadParameter("limit %v exceeds max iteration limit %v", limit, defaults.MaxIterationLimit)
 	}
+
 	// how many days of logs to search?
 	days := int(toUTC.Sub(fromUTC).Hours() / 24)
 	if days < 0 {
 		return nil, "", trace.BadParameter("invalid days")
 	}
-	filtered, err := l.matchingFiles(fromUTC, toUTC)
+	filesToSearch, err := l.matchingFiles(fromUTC, toUTC)
 	if err != nil {
 		return nil, "", trace.Wrap(err)
 	}
+
+	// If the start key is zero we don't need to perform any skipping.
+	// The "start" of the events we want to scan has already been found.
 	foundStart := startKey == ""
-	var total int
-	var lastKey string
-	// search within each file:
+
+	// lastReadKey contains the ID of the last read event ID.
+	var lastReadKey string
+
+	// newStartKey is the start key to be returned. By default we assume
+	// that all relevant events were returned in one query and therefore we initialize it to empty.
+	// If we do return early because we hit the provided iteration limit we set this to the lastReadKey.
+	var newStartKey string
+
 	dynamicEvents := make([]EventFields, 0)
-	for i := range filtered {
+
+	// search within each file:
+	for _, file := range filesToSearch {
 		var found []EventFields
-		found, lastKey, foundStart, err = l.findInFile(filtered[i].path, eventTypes, &total, limit, startKey, foundStart)
+		subLimit := 0
+		if limit > 0 {
+			subLimit = limit - len(dynamicEvents)
+		}
+
+		found, hitLimit, err := l.findInFile(file.path, eventTypes, subLimit)
 		if err != nil {
 			return nil, "", trace.Wrap(err)
 		}
-		dynamicEvents = append(dynamicEvents, found...)
-		if limit > 0 && total >= limit {
+
+		for _, foundEvent := range found {
+			eventID := foundEvent.GetString(EventID)
+
+			if !foundStart {
```

Comment:
startKey is technically startKey + 1 because this event is skipped. This is a bit confusing.

Reply:
Renamed to `startAfter`